File: bench/callback/callback.go
package main

import (
	"fmt"
	"time"
)

const (
	n       = 4096
	nIters  = 128
	nRuns   = 21
	nWarmup = 5
)

func mix(h, x uint32) uint32 { return (h ^ x) * 0x01000193 }

func medianUs(samples []float64) int {
	for i := 1; i < len(samples); i++ {
		v := samples[i]
		j := i - 1
		for j >= 0 && samples[j] > v {
			samples[j+1] = samples[j]
			j--
		}
		samples[j+1] = v
	}
	return int(samples[(len(samples)-1)>>1] * 1000)
}

func runKernel(a []float64, scale float64) uint32 {
	h := uint32(0x811c9dc5)
	b := make([]float64, len(a))
	for i := 0; i < nIters; i++ {
		iLocal := float64(i)
		for k := range a {
			b[k] = a[k]*scale + iLocal
		}
		for j := 0; j < len(a); j += 64 {
			h = mix(h, uint32(int32(b[j])))
		}
	}
	return h
}

func main() {
	a := make([]float64, n)
	for i := range a {
		a[i] = float64(i%97 - 48)
	}

	var cs uint32
	for i := 0; i < nWarmup; i++ {
		cs = runKernel(a, 2.0)
	}

	samples := make([]float64, nRuns)
	for i := 0; i < nRuns; i++ {
		t0 := time.Now()
		cs = runKernel(a, 2.0)
		samples[i] = float64(time.Since(t0).Nanoseconds()) / 1e6
	}
	fmt.Printf("median_us=%d checksum=%d samples=%d stages=%d runs=%d\n",
		medianUs(samples), cs, n*nIters, 1, nRuns)
}

File: bench/callback/callback.rs
use std::time::Instant;

const N: usize = 4096;
const N_ITERS: usize = 128;
const N_RUNS: usize = 21;
const N_WARMUP: usize = 5;

fn mix(h: u32, x: u32) -> u32 {
    (h ^ x).wrapping_mul(0x0100_0193)
}

fn median_us(samples: &mut [f64]) -> u64 {
    for i in 1..samples.len() {
        let v = samples[i];
        let mut j = i;
        while j > 0 && samples[j - 1] > v {
            samples[j] = samples[j - 1];
            j -= 1;
        }
        samples[j] = v;
    }
    (samples[(samples.len() - 1) >> 1] * 1000.0) as u64
}

fn run_kernel(a: &[f64], scale: f64) -> u32 {
    let mut h = 0x811c_9dc5u32;
    let mut b = vec![0.0f64; N];
    for i in 0..N_ITERS {
        let i_local = i as f64;
        for k in 0..N { b[k] = a[k] * scale + i_local; }
        for j in (0..N).step_by(64) {
            h = mix(h, b[j] as i32 as u32);
        }
    }
    h
}

fn main() {
    let a: Vec<f64> = (0..N).map(|i| (i % 97) as f64 - 48.0).collect();

    let mut cs = 0u32;
    for _ in 0..N_WARMUP { cs = run_kernel(&a, 2.0); }

    let mut samples = [0.0; N_RUNS];
    for sample in &mut samples {
        let t0 = Instant::now();
        cs = run_kernel(&a, 2.0);
        *sample = t0.elapsed().as_secs_f64() * 1000.0;
    }

    println!(
        "median_us={} checksum={} samples={} stages={} runs={}",
        median_us(&mut samples), cs, N * N_ITERS, 1, N_RUNS
    );
}
